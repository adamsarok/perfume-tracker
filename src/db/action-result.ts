interface ActionResult {
    error?: string;
    ok: boolean;
}