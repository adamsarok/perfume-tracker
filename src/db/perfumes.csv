Afnan;Supremacy in Oud;8,00
Ajmal;Evoke Midnight Edition Ajmal;8,50
Alhambra;Fabulo Intense;8,50
Alhambra;Hercules;8,00
Alhambra;Infini Oud Joyous;8,00
Alhambra;Kismet Angel;8,50
Alhambra;Porto Neroli;8,00
Alhambra;Russe Leather;8,00
Amouage;Reflection Man;8,00
Armaf;Derby Club House Gold;8,50
Boucheron;Ambre d'Alexandrie;8,00
Boucheron;Jaipur;8,00
CdG;Man 2;8,50
CH;Bad Boy Superstar;8,00
Chanel;Bleu;8,00
Chopard;1000 Miglia;8,00
Commodity;Milk+;8,00
Commodity;Moss+;8,50
Commodity;Whiskey;8,00
Davidoff;Zino;8,00
Diptyque;Eau Duelle EDP;9,00
ELDO;Putain de Palace;8,00
ELDO;Spice Must Flow;8,00
Fragrance World;Cocktail Intense;8,50
Fragrance World;Jasper Rouge;8,00
Fragrance World;Vanilla So Sweet;8,50
Givenchy;Reserve Privee EDP;8,50
Hermes;TdH EDT;8,50
Hermes;Terre d'Hermes Pure Perfume;8,50
Hermes;Voyage d'Hermes EDP;9,00
Histoires;7753 Mona Lisa;8,00
Initio;Blessed Baraka;8,00
Jo Malone;Myrrh & Tonka;8,00
JPG;Le Male Elixir;8,00
Lalique;Encre Noire;8,00
Lalique;Or Intemporel;8,50
Lolita Lempicka;Au Masculin;8,00
Lolita Lempicka;Au Masculin EDP Intense;9,00
Marabika;Tobacco Myth;8,00
Montblanc;Explorer;8,50
Mugler;A*Men;8,00
Mugler;A*Men Pure Havana;8,50
Mugler;A*Men Pure Malt;8,00
Mugler;Cologne;8,00
Narciso Rodriguez;Musc Noire Rose;8,00
Novellista;Velvet Dusk;8,50
Paco Rabane;1 million;8,00
Prada;Luna Rossa Black;8,00
Salvatore Ferragamo;F Black;8,00
Serge Lutens;Ambre Sultan;9,00
Serge Lutens;Chergui;8,00
Serge Lutens;L'eau Serge Lutens;8,50
The Merchant of Venice;Esperidi Water;8,50
Tom Ford;Noir;8,50
Tom Ford;Ombre Leather;9,00
Tom Ford;Vert d'Encens;8,00
Xerjoff;Naxos;8,00
YSL;Opium Pour Homme;8,00
Zara;Lisboa;8,50
Zara;Seoul;8,00
Zara;Tobacco Intense Dark;8,00
Zara;Tobacco Rich Warm;9,00
Zimaya;Sharaf Blend;7,50
Zoologist;Panda;8,00
Zoologist;Squid;8,50
Zoologist;Chipmunk;8,00
Zoologist;Civet;8,50
Zoologist;Cow;8,00
