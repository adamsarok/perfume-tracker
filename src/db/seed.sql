insert into "Perfume" (house, perfume, rating) VALUES ('Afnan','Supremacy in Oud',8.00);
insert into "Perfume" (house, perfume, rating) VALUES ('Afnan','Evoke Midnight Edition Ajmal',8.50);
insert into "Perfume" (house, perfume, rating) VALUES ('Alhambra','Fabulo Intense',8.50);