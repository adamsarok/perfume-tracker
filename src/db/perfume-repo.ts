'use server';

import db from ".";
import { Perfume, PerfumeTag, PerfumeWorn, Tag } from "@prisma/client";
import { revalidatePath } from "next/cache";
import { redirect } from "next/navigation";
import { z } from "zod";

const perfumeSchema = z.object({
    house: z.string().min(1),
    perfume: z.string().min(1),
    rating: z.string().min(1),
    notes: z.string().min(3),
    ml: z.string().min(1),
    // winter: z.string().nullable(),
    // summer: z.string().nullable(),
    // autumn: z.string().nullable(),
    // spring: z.string().nullable(),
})

interface UpdatePerfumeFormState {
    errors: {
        house?: string[];
        perfume?: string[];
        rating?: string[];
        notes?: string[];
        ml?: string[];
        winter?: string[];
        summer?: string[];
        autumn?: string[];
        spring?: string[];
        _form?: string[];
    },
    result: Perfume | null,
    state: 'init' | 'failed' | 'success'
}

export async function setImageObjectKey(id: number, objectKey: string) : Promise<ActionResult> {
    try {
        await db.perfume.update({
            where: {
                id
            }, 
            data: {
                imageObjectKey: objectKey
            }
        });
        return { ok: true };
    } catch (err: unknown) {
        if (err instanceof Error) {
            return {
                ok: false,
                error: err.message
            };
        } else {
            return { ok: false };
        }
    }
}

export async function upsertPerfume(id: number | null, tags: Tag[], formState: UpdatePerfumeFormState, formData: FormData)
    : Promise<UpdatePerfumeFormState> {
    try {
        const selectedTags = tags.map(x => x.id);
        const perf = perfumeSchema.safeParse({
            house: formData.get('house'),
            perfume: formData.get('perfume'),
            rating: formData.get('rating'),
            notes: formData.get('notes'),
            ml: formData.get('ml'),
        });
        if (!perf.success) {
            return {
                errors: perf.error.flatten().fieldErrors,
                result: null,
                state: 'failed'
            }
        }
        if (!parseFloat(perf.data.rating)) {
            return {
                errors: { rating: ['Not a valid rating'] },
                result: null,
                state: 'failed'
            }
        }
        const summmer = formData.get('summer') ? true : false;
        const winter = formData.get('winter') ? true : false;
        const spring = formData.get('spring') ? true : false;
        const autumn = formData.get('autumn') ? true : false;
        if (id) {
            const currentTags = await db.perfumeTag.findMany({
                where: {
                    perfumeId: id
                }
            });
            const currentTagIds = currentTags.map(x => x.tagId);
            const tagIdsToRemove = currentTags
                .filter(x => !selectedTags.includes(x.tagId))
                .map(m => m.id);
            const tagsToAdd = selectedTags.filter(x => !currentTagIds.includes(x));
            const result = await db.perfume.update({
                where: {
                    id: id
                },
                data: {
                    house: perf.data.house,
                    perfume: perf.data.perfume,
                    rating: parseFloat(perf.data.rating),
                    notes: perf.data.notes,
                    ml: parseInt(perf.data.ml),
                    summer: summmer,
                    winter: winter,
                    spring: spring,
                    autumn: autumn,
                    tags: {
                        createMany: {
                            data: tagsToAdd.map(x => ({
                                tagId: x
                            }))
                        },
                        deleteMany: {
                            id: {
                                in: tagIdsToRemove
                            }
                        }
                    }
                }
            });
            if (result) return  {
                errors: {},
                result,
                state: 'success'
            }
        } else {
            const result = await db.perfume.create({
                data: {
                    house: perf.data.house,
                    perfume: perf.data.perfume,
                    rating: parseFloat(perf.data.rating),
                    notes: perf.data.notes,
                    ml: parseInt(perf.data.ml),
                    summer: summmer,
                    winter: winter,
                    spring: spring,
                    autumn: autumn,
                    tags: {
                        createMany: {
                            data: tags.map(t => ({
                                tagId: t.id
                            }))
                        }
                    }
                }
            });
            if (result) return  {
                errors: {},
                result,
                state: 'success'
            }
        }
    } catch (err: unknown) {
        if (err instanceof Error) {
            return {
                errors: { _form: [err.message] },
                result: null,
                state: 'failed'
            };
        }
    }
    return {
        errors: { _form: ['Unknown error occured'] },
        result: null,
        state: 'failed'
    };
}

export async function getPerfumesForSelector(): Promise<Perfume[]> {
    return await db.perfume.findMany({
        orderBy: [
            {
                house: 'asc',
            },
            {
                perfume: 'asc',
            },
        ]
    });
}


export async function getPerfumeWithTags(id: number) {
    return await db.perfume.findFirst({
        where: {
            id: id
        },
        include: {
            tags: {
                include: {
                    tag: true
                }
            }
        }
    });
}

export interface PerfumeWithTagDTO {
    perfume: Perfume,
    tags: Tag[]
}

export async function getPerfumesWithTags() : Promise<PerfumeWithTagDTO[]> {
    const perfumes = await db.perfume.findMany({
        include: {
            tags: {
                include: {
                    tag: true
                }
            }
        }
    });
    const result: PerfumeWithTagDTO[] = [];
    perfumes.map((p) => {
        result.push({
            perfume: p,
            tags: p.tags.map((t) => {
                return {
                    id: t.tag.id,
                    color: t.tag.color,
                    tag: t.tag.tag
                }
            })
        });
    });
    return result;
}

export async function deletePerfume(id: number) : Promise<{ error: string | null }> {
    try {
        await db.perfume.delete({
            where: {
                id
            }
        });
    } catch (err: unknown) {
        if (err instanceof Error) {
            return {
                error: err.message
            };
        } else {
            return {
                error: 'Unknown error occured'
            };
        }
    }
    return {
        error: null
    };
}
