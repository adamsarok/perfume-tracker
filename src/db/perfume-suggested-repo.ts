'use server';

import db from ".";

export async function getAlreadySuggestedPerfumeIds(dayFilter: Date) {
    return await db.perfumeSuggested.findMany({
        where: {
            suggestedOn: {
                gte: dayFilter,
            },
        },
        select: {
            perfumeId: true,
        },
        distinct: ["perfumeId"],
    });
}


export async function addSuggested(perfumeId: number) {
    await db.perfumeSuggested.create({
        data: {
            suggestedOn: new Date(),
            perfume: {
                connect: {
                    id: perfumeId
                }
            }
        },
    });
}