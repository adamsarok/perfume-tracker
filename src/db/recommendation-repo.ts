'use server';

import { Recommendation } from "@prisma/client";
import db from ".";

export async function insertRecommendation(query: string, recommendations: string | null) : Promise<ActionResult> {
    try {
        if (!query) return { ok: false, error: "Query empty" };
        await db.recommendation.create({
            data: {
                query,
                recommendations: recommendations ?? ''
            }
        });
        return { ok: true };
    } catch (err: unknown) {
        if (err instanceof Error) {
            return {
                ok: false,
                error: err.message
            };
        } else {
            return {
                ok: false,
                error: 'Unknown error occured'
            };
        }
    }
}

export async function getRecommendations() : Promise<Recommendation[]> {
    return await db.recommendation.findMany();
}


