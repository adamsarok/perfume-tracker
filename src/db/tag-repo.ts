'use server';

import { Tag } from "@prisma/client";
import db from ".";
import { z } from "zod";

interface InsertTagFormState {
    result: {
        id: number,
        tag: string,
        color: string
    } | null
    errors: {
        name?: string[];
        color?: string[];
        _form?: string[];
    }
}

const tagSchema = z.object({
    tag: z.string().min(1),
    color: z.string().min(1) //TODO color validate
})


export async function insertTag(formState: InsertTagFormState, formData: FormData) : Promise<InsertTagFormState> {
    try {
        console.log(formData);
        const perf = tagSchema.safeParse({
            tag: formData.get('tag'),
            color: formData.get('color')
        });
        if (!perf.success) {
            console.log(perf.error.flatten().fieldErrors);
            return {
                result: null,
                errors: perf.error.flatten().fieldErrors,
            }
        }
        const result = await db.tag.create({
            data: {
                tag: perf.data.tag,
                color: perf.data.color,
            }
        });
        console.log(result);
        return {
            result: {
                id: result.id,
                tag: result.tag,
                color: result.color
            },
            errors: { _form: [] }
        };
    } catch (err: unknown) {
        if (err instanceof Error) {
            return {
                result: null,
                errors: { _form: [err.message] }
            };
        } else {
            return {
                result: null,
                errors: { _form: ['Unknown error occured'] }
            };
        }
    }
}

export async function getTags() : Promise<Tag[]> {
    return await db.tag.findMany();
}

interface TagResult {
    success: boolean;
    error: string;
}

export async function deleteTag(id: number) : Promise<TagResult> {
    try {    
        await db.tag.delete({
                where: {
                    id
                }
            });
        return { success: true, error: "" }
    } catch (error: any) {
        return { success: false, error: error.message || 'Failed to delete tag' };
    }
}

export async function updateTag(id: number, tag: string, color: string) : Promise<TagResult> {
    try {    
        await db.tag.update({
                where: {
                    id
                },
                data: {
                    tag,
                    color
                }
            });
        return { success: true, error: "" }
    } catch (error: any) {
        return { success: false, error: error.message || 'Failed to delete tag' };
    }
}


