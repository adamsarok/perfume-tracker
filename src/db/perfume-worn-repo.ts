'use server';

import { Perfume, PerfumeWorn, Tag } from "@prisma/client";
import { revalidatePath } from "next/cache";
import db from ".";
import * as perfumeRepo from "./perfume-repo"

export async function GetWorn() {
    return await db.perfumeWorn.findMany({
        include: {
            perfume: true
        },
        orderBy: [
            {
                wornOn: 'desc',
            },
        ]
    });
}

export async function DeleteWear(id: number) {
    if (!id) return;
    const idNum: number = parseInt(id.toString());
    await db.perfumeWorn.delete({
        where: {
            id
        },
    });
    revalidatePath('/');
}

//warning todo utc
export async function WearPerfume(id: number) {
    if (!id) return;
    const idNum: number = parseInt(id.toString());
    const today = new Date();
    today.setHours(0, 0, 0);
    const tomorrow = new Date();
    tomorrow.setDate(today.getDate() + 1);
    const alreadyWornToday = await db.perfumeWorn.findFirst({
        where: {
            perfumeId: idNum,
            wornOn: {
                gte: today,
                lt: tomorrow
            }
        }
    });
    if (alreadyWornToday) {
        console.log("Already wearing this perfume today");
        return;
    }
    await db.perfumeWorn.create({
        data: {
            wornOn: new Date(),
            perfume: {
                connect: {
                    id: idNum
                }
            }
        },
    });
    revalidatePath('/');
}

export interface PerfumeWornDTO {
    perfume: Perfume,
    wornTimes: number | undefined,
    lastWorn: Date | undefined,
    tags: Tag[]
}

export async function GetAllPerfumesWithWearCount(): Promise<PerfumeWornDTO[]> {
    const worn = await db.perfumeWorn.groupBy({
        by: ['perfumeId'],
        _count: {
            id: true
        },
        _max: {
            wornOn: true
        }
    });
    const perfumes = await perfumeRepo.GetPerfumesWithTags();
    let m = new Map();
    perfumes.forEach(function (x) {
        let dto: PerfumeWornDTO = {
            perfume: x,
            wornTimes: undefined,
            lastWorn: undefined,
            tags: x.tags.map(x => x.tag)
        }
        m.set(x.id, dto);
    });
    worn.forEach(function (x) {
        let dto = m.get(x.perfumeId);
        dto.wornTimes = x._count.id;
        dto.lastWorn = x._max.wornOn;
    })
    return Array.from(m.values());
}