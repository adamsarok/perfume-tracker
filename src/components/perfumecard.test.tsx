import { expect, test } from 'vitest'
import { render } from '@testing-library/react'
import PerfumeCard from '../components/perfumecard'
import { WornWithPerfume } from '@/db/perfume-worn-repo';

test('PerfumeCard snapshot', () => {
    // Snapshot test, if I ever need one?
    // const mockWorn: WornWithPerfume =  {
    //     id: 1,
    //     perfumeId: 1,
    //     wornOn: new Date('2024-03-15'),
    //     perfume: {
    //         id: 1, house: 'Dior', perfume: 'Sauvage', rating: 9, notes: 'Fresh', ml: 100,
    //         imageObjectKey: ''
    //     },
    //     tags: [{ id: 1, tag: 'Fresh', color: '#00ff00' }]
    // };

    // const { asFragment } = render(<PerfumeCard worn={mockWorn} />);
    // expect(asFragment()).toMatchSnapshot();
});