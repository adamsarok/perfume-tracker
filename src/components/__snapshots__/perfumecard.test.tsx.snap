// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PerfumeCard snapshot 1`] = `
<DocumentFragment>
  <form>
    <div
      class="flex flex-col relative overflow-hidden height-auto text-foreground box-border bg-content1 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 shadow-medium rounded-large transition-transform-background motion-reduce:transition-none min-w-96"
      tabindex="-1"
    >
      <div
        class="flex p-3 z-10 w-full justify-start items-center shrink-0 overflow-inherit color-inherit subpixel-antialiased rounded-t-large"
      >
        <a
          class="relative inline-flex items-center tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-medium text-primary no-underline hover:opacity-80 active:opacity-disabled transition-opacity"
          href="/perfumes/1/"
          role="link"
          tabindex="0"
        >
          <span
            class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-default text-default-foreground rounded-full semi-bold"
            tabindex="-1"
          >
            <span
              aria-label="SA"
              class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 font-normal text-center text-inherit"
              role="img"
            >
              SA
            </span>
          </span>
          <div
            class="text-small leading-none text-default-600 ml-4"
          >
            Dior - Sauvage
          </div>
          <span
            class="flex relative justify-center items-center box-border overflow-hidden align-middle z-0 outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 w-10 h-10 text-tiny bg-secondary text-secondary-foreground rounded-full semi-bold ml-4"
            tabindex="-1"
          >
            <span
              aria-label="9"
              class="absolute top-1/2 left-1/2 -translate-x-1/2 -translate-y-1/2 font-normal text-center text-inherit"
              role="img"
            >
              9
            </span>
          </span>
        </a>
        <button
          class="z-0 group inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-unit-3 min-w-unit-16 h-unit-8 text-tiny gap-unit-2 rounded-small [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none bg-danger text-danger-foreground data-[hover=true]:opacity-hover absolute right-5 max-w-4"
          type="button"
        >
          X
        </button>
      </div>
      <div
        class="relative flex w-full p-3 flex-auto flex-col place-content-inherit align-items-inherit h-auto break-words text-left overflow-y-auto subpixel-antialiased"
      >
        <p
          class="text-small tracking-tight text-default-400"
        >
          Worn on: Fri Mar 15 2024
        </p>
      </div>
    </div>
  </form>
</DocumentFragment>
`;
